File: problems/CSES/Graph_Algorithms/cycle_finding/cycle_finding.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())

    edges = []

    for _ in range(m):
        u, v, w = map(int, input().split()) # From / To / Cost

        u -= 1
        v -= 1

        edges.append((u, v, w))

    # Bellman-Ford
    dist = [0] * n
    parent = [-1] * n

    x = -1

    for _ in range(n):
        x = -1

        for u, v, w in edges:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                x = v

    if x == -1:
        print("NO")
        return

    for _ in range(n):
        x = parent[x]

    cycle = [x]
    cur = parent[x]

    while cur != x:
        cycle.append(cur)
        cur = parent[cur]

    cycle.append(x)

    cycle.reverse()

    print("YES")
    print(*(v + 1 for v in cycle))

main()

File: problems/CSES/Graph_Algorithms/cycle_finding/cycle_finding.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

struct Edge {
    int u, v;
    ll w;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<Edge> edges;

    for (int i = 0; i < m; i++) {
        int u, v;
        ll w;

        cin >> u >> v >> w; // From / To / Cost

        u--;
        v--;

        edges.push_back({u, v, w});
    }

    // Bellman-Ford
    vector<ll> dist(n, 0);
    vector<int> parent(n, -1);

    int x = -1;

    for (int i = 0; i < n; i++) {
        x = -1;

        for (auto [u, v, w] : edges) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                parent[v] = u;
                x = v;
            }
        }
    }

    // No negative cycle
    if (x == -1) {
        cout << "NO\n";
        return 0;
    }

    // Move inside the negative cycle
    for (int i = 0; i < n; i++) {
        x = parent[x];
    }

    // Reconstruct cycle
    vector<int> cycle;

    int cur = x;

    do {
        cycle.push_back(cur);
        cur = parent[cur];
    } while (cur != x);

    cycle.push_back(x);

    reverse(cycle.begin(), cycle.end());

    cout << "YES\n";

    for (int v : cycle) {
        cout << v + 1 << ' ';
    }

    cout << '\n';

    return 0;
}
